string_grouper_utils/string_grouper_utils.py

```diff
@@ -0,0 +1,203 @@
+
+
+def group_rep_transform(method: str,
+                        weights: pd.Series,
+                        grouped_data,
+                        group_col,
+                        record_id_col,
+                        record_name_col) -> Union[pd.Series, pd.DataFrame]:
+    group_of_master_id = get_column(group_col, grouped_data)
+    group_of_master_id = group_of_master_id.rename('raw_group_id').reset_index().rename(columns={'index': 'weight'})
+    group_of_master_id['weight'] = weights
+    group_of_master_id['group_rep'] = \
+        group_of_master_id.groupby('raw_group_id', sort=False)['weight'].transform(method)
+    record_id_col = get_column(record_id_col, grouped_data)
+    new_rep = record_id_col.iloc[group_of_master_id.group_rep].reset_index(drop=True).rename(None)
+    if record_name_col is None:
+        return new_rep
+    else:
+        record_name_col = get_column(record_name_col, grouped_data)
+        new_rep_name = record_name_col.iloc[group_of_master_id.group_rep].reset_index(drop=True).rename(None)
+        return pd.concat([new_rep, new_rep_name], axis=1)
+
+
+def get_column(col: Union[str, int, List[Union[str, int]]], data: pd.DataFrame):
+    if isinstance(col, str):
+        return data.loc[:, col]
+    elif isinstance(col, int):
+        return data.iloc[:, col]
+    elif isinstance(col, List):
+        return pd.concat([get_column(m, data) for m in col], axis=1)
+
+
+def parse_timestamps(timestamps: pd.Series, **kwargs) -> pd.Series:
+    error_msg = f"timestamps must be a Series of date-like or datetime-like strings"
+    error_msg += f" or datetime datatype or pandas Timestamp datatype or numbers"
+    if is_series_of_strings(timestamps):
+        # if any of the strings is not datetime-like raise an exception
+        if timestamps.to_frame().applymap(is_not_date).squeeze().any():
+            raise Exception(error_msg)
+        else:
+            # convert strings to numpy datetime64
+            return timestamps.copy().transform(lambda x: np.datetime64(parse(x, **kwargs)))
+    elif is_series_of_Timestamps(timestamps):
+        # convert pandas Timestamps to numpy datetime64
+        return timestamps.copy().transform(lambda x: x.to_numpy())
+    elif is_series_of_datetimes(timestamps):
+        # convert python datetimes to numpy datetime64
+        return timestamps.copy().transform(lambda x: np.datetime64(x))
+    elif not is_series_of_numbers(timestamps):
+        raise Exception(error_msg)
+    return timestamps
+
+
+def is_not_date(string, fuzzy=True):
+    """
+    Return whether the string can be interpreted as a date.
+    :param string: str, string to check for date
+    :param fuzzy: bool, ignore unknown tokens in string if True
+    """
+    try:
+        parse(string, fuzzy=fuzzy)
+        return False
+    except ValueError:
+        return True
+
+
+def is_series_of_strings(series_to_test: pd.Series) -> bool:
+    if series_to_test.to_frame().applymap(
```

Comment:
We have three times almost the same function but with a different type (repetition). Maybe create one extra function "_is_series_of_type(serie, type: str)" The functions just return this function with a different type?

Reply:
Good catch!  I'm surprised my IDE didn't catch this for me.